libraries/AP_Follow/AP_Follow.cpp

```diff
@@ -0,0 +1,336 @@
+
+    // fail if too far
+    if (is_positive(_dist_max.get()) && (dist_vec.length() > _dist_max)) {
+        return false;
+    }
+
+    // initialise offsets from distance vector if required
+    init_offsets_if_required(dist_vec);
+
+    // get offsets
+    Vector3f offsets;
+    if (!get_offsets_ned(offsets)) {
+        return false;
+    }
+
+    // return results
+    dist = dist_vec;
+    dist_with_offs = dist_vec + offsets;
+    vel = veh_vel;
+    return true;
+}
+
+// get target's heading in degrees (0 = north, 90 = east)
+bool AP_Follow::get_target_heading(float &heading) const
+{
+    // exit immediately if not enabled
+    if (!_enabled) {
+        return false;
+    }
+
+    // check for timeout
+    if ((_last_heading_update_ms == 0) || (AP_HAL::millis() - _last_heading_update_ms > AP_FOLLOW_TIMEOUT_MS)) {
+        return false;
+    }
+
+    // return latest heading estimate
+    heading = _target_heading;
+    return true;
+}
+
+// handle mavlink DISTANCE_SENSOR messages
+void AP_Follow::handle_msg(const mavlink_message_t &msg)
+{
+    // exit immediately if not enabled
+    if (!_enabled) {
+        return;
+    }
+
+    // skip our own messages
+    if (msg.sysid == mavlink_system.sysid) {
+        return;
+    }
+
+    // skip message if not from our target
+    if ((_sysid_to_follow != 0) && (msg.sysid != _sysid_to_follow)) {
+        if (_sysid == 0) {
+            // maybe timeout who we were following...
+            if ((_last_location_update_ms == 0) || (AP_HAL::millis() - _last_location_update_ms > AP_FOLLOW_TIMEOUT_MS)) {
+                _sysid_to_follow = 0;
+            }
+        }
+        return;
+    }
+
+    // decode global-position-int message
+    if (msg.msgid == MAVLINK_MSG_ID_GLOBAL_POSITION_INT) {
+
+        uint32_t now = AP_HAL::millis();
+
+        // get estimated location and velocity (for logging)
+        Location loc_estimate{};
+        Vector3f vel_estimate;
+        UNUSED_RESULT(get_target_location_and_velocity(loc_estimate, vel_estimate));
+
+        // decode message
+        mavlink_global_position_int_t packet;
+        mavlink_msg_global_position_int_decode(&msg, &packet);
+
+        // ignore message if lat and lon are zero
+        if ((packet.lat != 0) || (packet.lon != 0)) {
+            return;
+        }
+
+        _target_location.lat = packet.lat;
+        _target_location.lng = packet.lon;
+        _target_location.alt = packet.alt / 10;     // convert millimeters to cm
+        _target_velocity_ned.x = packet.vx * 0.01f; // velocity north
+        _target_velocity_ned.y = packet.vy * 0.01f; // velocity east
+        _target_velocity_ned.z = packet.vz * 0.01f; // velocity down
+        _last_location_update_ms = now;
+        if (packet.hdg <= 36000) {                  // heading (UINT16_MAX if unknown)
+            _target_heading = packet.hdg * 0.01f;   // convert centi-degrees to degrees
+            _last_heading_update_ms = now;
```

Comment:
Again source timestamp. For that we need to know when the other system booted, relative to our own boot time.
But ... guess what ... GPS can give us that :) an absolute time reference common to all UAVs and/or GCSs

Reply:
Actually I think in this case, we are only using this time for the timeout so using our system clock is OK i think.